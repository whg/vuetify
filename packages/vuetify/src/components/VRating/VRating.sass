@import './_variables.scss'

.v-rating
  max-width: 100%
  white-space: nowrap

  .v-icon
    +ltr()
      transform: scaleX(1)

    +rtl()
      transform: scaleX(-1)

  &--readonly
    pointer-events: none

.v-rating__item
  display: inline-flex
  flex-direction: column
  align-items: center

  &--bottom
    flex-direction: column-reverse
