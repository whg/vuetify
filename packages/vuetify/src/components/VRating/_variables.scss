$rating-white-space: nowrap !default;
$rating-item-align-items: center !default;
