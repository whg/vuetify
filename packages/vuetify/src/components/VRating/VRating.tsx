// Styles
import './VRating.sass'

// Components
import { VBtn } from '../VBtn'

// Composables
import { useProxiedModel } from '@/composables/proxiedModel'
import { makeDensityProps } from '@/composables/density'
import { makeSizeProps } from '@/composables/size'
import { makeTagProps } from '@/composables/tag'
import { useRefs } from '@/composables/refs'

// Utilities
import { computed, defineComponent, nextTick, ref } from 'vue'
import { createRange, keyCodes, makeProps } from '@/util'

// Types
import type { ComponentPublicInstance, Prop, Ref } from 'vue'

export default defineComponent({
  name: 'VRating',

  props: makeProps({
    backgroundColor: {
      type: String,
      default: 'accent',
    },
    color: {
      type: String,
      default: 'primary',
    },
    clearable: Boolean,
    emptyIcon: {
      type: String,
      default: '$ratingEmpty',
    },
    fullIcon: {
      type: String,
      default: '$ratingFull',
    },
    halfIcon: {
      type: String,
      default: '$ratingHalf',
    },
    halfIncrements: Boolean,
    hover: Boolean,
    length: {
      type: [Number, String],
      default: 5,
    },
    readonly: Boolean,
    modelValue: {
      type: Number,
      default: 0,
    },
    ariaLabel: {
      type: String,
      default: '$vuetify.rating.ariaLabel',
    },
    disabled: Boolean,
    itemLabels: Array as Prop<string[]>,
    itemLabelPosition: {
      type: String,
      default: 'top',
      validator: (v: any) => ['top', 'bottom'].includes(v),
    },
    ripple: Boolean,
    ...makeDensityProps(),
    ...makeSizeProps(),
    ...makeTagProps(),
  }),

  setup (props, { slots }) {
    const rating = useProxiedModel(props, 'modelValue') as any as Ref<number> // TODO: Why is type not working?
    const length = computed(() => Number(props.length))
    const hoverIndex = ref(-1)

    const icons = computed(() => {
      const isHalfEvent = (e: MouseEvent): boolean => {
        const rect = e.target && (e.target as HTMLElement).getBoundingClientRect()
        const isHalf = !!rect && (e.pageX - rect.left) < rect.width / 2

        // TODO: handle rtl
        // return isRtl ? !isHalf : isHalf
        return isHalf
      }

      const genHoverIndex = (e: MouseEvent, i: number) => {
        const isHalf = props.halfIncrements && isHalfEvent(e)

        return i + (isHalf ? 0.5 : 1)
      }

      const createSlotProps = (index: number) => {
        const isHovering = props.hover && hoverIndex.value > -1
        const isFilled = Math.floor(rating.value) > index
        const isHovered = Math.floor(hoverIndex.value) > index
        const isHalfHovered = !isHovered && (hoverIndex.value - index) % 1 > 0
        const isHalfFilled = !isFilled && (rating.value - index) % 1 > 0

        const isFullIcon = isHovering ? isHovered : isFilled
        const isHalfIcon = isHovering ? isHalfHovered : isHalfFilled

        const icon = isFullIcon ? props.fullIcon
          : isHalfIcon ? props.halfIcon
          : props.emptyIcon

        const onMouseenter = (e: MouseEvent): void => {
          hoverIndex.value = genHoverIndex(e, index)
        }

        const onMouseleave = (): void => {
          hoverIndex.value = -1
        }

        return {
          onMouseenter: props.hover ? onMouseenter : undefined,
          onMouseleave: props.hover ? onMouseleave : undefined,
          onMousemove: props.hover && props.halfIncrements ? onMouseenter : undefined,
          onClick: (e: MouseEvent) => {
            if (props.readonly) return

            // If detail === 0 then click is triggered by keyboard
            if (e.detail === 0) {
              const currentIndex = Math.floor(rating.value)
              if (rating.value - 1 === index && props.clearable) {
                rating.value = 0
              } else if (currentIndex !== index || !props.halfIncrements) {
                rating.value = index + (props.halfIncrements ? 0.5 : 1)
              } else {
                rating.value += 0.5
              }
            } else {
              let newValue = genHoverIndex(e, index)
              if (props.clearable && rating.value === newValue) {
                rating.value = 0
              } else {
                rating.value = newValue
              }
            }
          },
          icon,
          color: isFilled || isHalfFilled || isHovered ? props.color : props.backgroundColor,
          size: props.size,
          disabled: props.disabled,
          // TODO: fix when locale is done
          // ariaLabel: this.$vuetify.lang.t(props.iconLabel, index + 1, length.value)
          ariaLabel: String(index + 1),
          isFilled,
          isHovered,
          isHalfHovered,
          isHalfFilled,
          index,
          value: rating.value,
          hasLabels: !!props.itemLabels?.length || !!slots['item-label'],
          label: props.itemLabels && props.itemLabels[index],
          ripple: props.ripple,
          density: props.density,
          readonly: props.readonly,
          tabindex: props.readonly ? -1 : undefined
        }
      }

      return createRange(length.value).map(i => createSlotProps(i))
    })

    const { refs, updateRef } = useRefs<ComponentPublicInstance>()

    const onKeydown = (e: KeyboardEvent) => {
      const increment = props.halfIncrements ? 0.5 : 1
      if (e.keyCode === keyCodes.left && rating.value > 0) {
        rating.value -= increment
        nextTick(() => refs.value[Math.floor(rating.value)]?.$el.focus())
      } else if (e.keyCode === keyCodes.right && rating.value < length.value) {
        rating.value += increment
        nextTick(() => refs.value[Math.floor(rating.value - 0.5)]?.$el.focus())
      }
    }

    return () => (
      <props.tag
        class={[
          "v-rating",
          {
            'v-rating--readonly': props.readonly,
          }
        ]}
        onKeydown={onKeydown}
      >
        {icons.value.map(iconProps => (
          <div
            key={iconProps.index}
            class={[
              "v-rating__item",
              {
                'v-rating__item--bottom': props.itemLabelPosition === 'bottom',
              }
            ]}
          >
            {
              !iconProps.hasLabels ? undefined
              : slots['item-label'] ? slots['item-label'](iconProps)
              : iconProps.label ? <span>{iconProps.label}</span>
              : <span>&nbsp;</span>
            }
            {slots.item ? slots.item(iconProps) : (
              <VBtn
                ref={(e: any) => updateRef(e, iconProps.index)}
                color={iconProps.color}
                ripple={iconProps.ripple}
                size={iconProps.size}
                icon={iconProps.icon}
                onClick={iconProps.onClick}
                onMouseenter={iconProps.onMouseenter}
                onMouseleave={iconProps.onMouseleave}
                onMousemove={iconProps.onMousemove}
                aria-label={iconProps.ariaLabel}
                disabled={iconProps.disabled}
                density={iconProps.density}
                tabindex={iconProps.tabindex}
              />
            )}
          </div>
        ))}
      </props.tag>
    )
  }
})
