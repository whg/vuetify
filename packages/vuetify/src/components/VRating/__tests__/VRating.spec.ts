// Components
import VRating from '../VRating'

// Utilities
import { mount } from '@vue/test-utils'
import { createVuetify } from '@/framework'
import { h } from '@vue/runtime-core'

describe('VRating.ts', () => {
  const vuetify = createVuetify()
  const mountFunction = (options?: any) => {
    return mount(VRating, {
      ...options,
      global: {
        plugins: [vuetify],
      },
    })
  }

  it('should respond to prop value changes', async () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ modelValue: 1 })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should respond to user interaction', async () => {
    const wrapper = mountFunction()

    const items = wrapper.findAll('.v-rating__item > button')

    await items[0].trigger('click')

    expect(wrapper.emitted('update:modelValue')).toEqual([
      [1],
    ])
  })

  it('should not clear value if using clearable prop', async () => {
    const wrapper = mountFunction({
      props: {
        modelValue: 1,
        clearable: true,
      },
    })

    const items = wrapper.findAll('.v-rating__item button')

    await items[0].trigger('click')

    expect(wrapper.emitted('update:modelValue')).toEqual([
      [0],
    ])

    await wrapper.setProps({ modelValue: 1, clearable: false })

    await items[0].trigger('click')

    expect(wrapper.emitted('update:modelValue')).toEqual([
      [0],
      [1],
    ])
  })

  it('should not react to events when readonly', async () => {
    const wrapper = mountFunction({
      props: {
        readonly: true,
      },
    })

    const items = wrapper.findAll('.v-rating__item button')

    await items[0].trigger('click')

    expect(wrapper.emitted('update:modelValue')).toBeUndefined()
  })

  it('should change icon on hover', async () => {
    const wrapper = mountFunction({
      props: {
        hover: true,
      },
    })

    const items = wrapper.findAll('.v-rating__item > button')

    await items[2].trigger('mouseenter')

    expect(wrapper.html()).toMatchSnapshot()
  })

  // it('should check for half event', () => {
  //   const wrapper = mountFunction()

  //   const event = new MouseEvent('hover')
  //   expect(wrapper.vm.genHoverIndex(event, 1)).toBe(2)

  //   wrapper.setProps({ halfIncrements: true })

  //   expect(wrapper.vm.genHoverIndex({
  //     pageX: 0,
  //     target: {
  //       getBoundingClientRect: () => ({ width: 10, left: 0 }),
  //     },
  //   }, 1)).toBe(1.5)

  //   expect(wrapper.vm.genHoverIndex({
  //     pageX: 6,
  //     target: {
  //       getBoundingClientRect: () => ({ width: 10, left: 0 }),
  //     },
  //   }, 1)).toBe(2)
  // })

  // it('should check for half event in rtl', () => {
  //   const wrapper = mountFunction({
  //     propsData: { halfIncrements: true },
  //     mocks: {
  //       $vuetify: {
  //         rtl: true,
  //         lang: {
  //           t: str => str,
  //         },
  //       },
  //     },
  //   })

  //   const event = new MouseEvent('hover')
  //   expect(wrapper.vm.genHoverIndex(event, 1)).toBe(1.5)

  //   wrapper.setProps({ halfIncrements: true })

  //   expect(wrapper.vm.genHoverIndex({
  //     pageX: 0,
  //     target: {
  //       getBoundingClientRect: () => ({ width: 10, left: 0 }),
  //     },
  //   }, 1)).toBe(2)

  //   expect(wrapper.vm.genHoverIndex({
  //     pageX: 6,
  //     target: {
  //       getBoundingClientRect: () => ({ width: 10, left: 0 }),
  //     },
  //   }, 1)).toBe(1.5)
  // })

  it('should render a scoped item slot', () => {
    const wrapper = mountFunction({
      slots: {
        item: (props: any) => h('div', props.index),
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a scoped item-label slot', () => {
    const wrapper = mountFunction({
      slots: {
        'item-label': () => h('span', ['foo']),
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  // it('should bind mousemove listener', () => {
  //   const onMouseEnter = jest.fn()
  //   const wrapper = mountFunction({
  //     propsData: {
  //       halfIncrements: true,
  //       hover: true,
  //     },
  //     methods: { onMouseEnter },
  //   })

  //   const icon = wrapper.find('.v-icon')

  //   icon.trigger('mousemove')

  //   expect(onMouseEnter).toHaveBeenCalled()
  // })

  // it('should reset hoverIndex on mouse leave', () => {
  //   jest.useFakeTimers()
  //   const wrapper = mountFunction({
  //     propsData: { hover: true },
  //   })

  //   const icon = wrapper.find('.v-icon')

  //   expect(wrapper.vm.hoverIndex).toBe(-1)

  //   icon.trigger('mouseenter')
  //   jest.runAllTimers()

  //   expect(wrapper.vm.hoverIndex).toBe(1)

  //   icon.trigger('mouseleave')
  //   jest.runAllTimers()

  //   expect(wrapper.vm.hoverIndex).toBe(-1)
  // })
})
