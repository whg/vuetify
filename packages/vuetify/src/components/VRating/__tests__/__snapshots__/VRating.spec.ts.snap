// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VRating.ts should change icon on hover 1`] = `
<div class="v-rating">
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-primary v-btn--density-default v-btn--size-default"
            aria-label="1"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-primary v-btn--density-default v-btn--size-default"
            aria-label="2"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-primary v-btn--density-default v-btn--size-default"
            aria-label="3"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="4"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="5"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
</div>
`;

exports[`VRating.ts should render a scoped item slot 1`] = `
<div class="v-rating">
  <div class="v-rating__item">
    <div>
      0
    </div>
  </div>
  <div class="v-rating__item">
    <div>
      1
    </div>
  </div>
  <div class="v-rating__item">
    <div>
      2
    </div>
  </div>
  <div class="v-rating__item">
    <div>
      3
    </div>
  </div>
  <div class="v-rating__item">
    <div>
      4
    </div>
  </div>
</div>
`;

exports[`VRating.ts should render a scoped item-label slot 1`] = `
<div class="v-rating">
  <div class="v-rating__item">
    <span>
      foo
    </span>
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="1"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <span>
      foo
    </span>
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="2"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <span>
      foo
    </span>
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="3"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <span>
      foo
    </span>
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="4"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <span>
      foo
    </span>
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="5"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
</div>
`;

exports[`VRating.ts should respond to prop value changes 1`] = `
<div class="v-rating">
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="1"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="2"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="3"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="4"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="5"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
</div>
`;

exports[`VRating.ts should respond to prop value changes 2`] = `
<div class="v-rating">
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-primary v-btn--density-default v-btn--size-default"
            aria-label="1"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="2"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="3"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="4"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
  <div class="v-rating__item">
    <button type="button"
            class="v-btn v-btn--icon v-theme--light text-accent v-btn--density-default v-btn--size-default"
            aria-label="5"
    >
      <span class="v-btn__overlay">
      </span>
      <i class="mdi-star-outline mdi v-icon notranslate v-icon--size-default"
         aria-hidden="true"
      >
      </i>
    </button>
  </div>
</div>
`;
